### ISSUE TITLE
Address Matching -- deal with address ranges

### ISSUE THREAD

**Issue by RDmitchell** [SEED-platform/seed#317]
The address parsing module has some issues with address ranges. If there is no whitespace in an address range it parses the 'house' appropriately (though not necessarily in a useful way - there is no indication of range start and range end). If there is whitespace then the parser is less predictable and the output is not so well formed.

Address ranges and matching datasets on their permutations have been a real struggle for me so this might add unwanted complexity to your normalization task but getting it right could improve matching.

example test addresses

322 S Green St #house:322
300-22 S Green St #house:300-22
300/322 S Green St #house:300/322
300 322 S Green St #house:300,street_name:'322 S green'

**Comment by RDmitchell** [comment]
@nicholasserra -- can you see if this is still an issue, and if it is, propose a way to deal with the address ranges without whitespaces. I created the issue above, but I believe it was emailed to me by another developer. 

If it is useful, there is a set of data associated with this issue that has several records with address ranges. In general they are well formed, although one does not have white space between the address ranges. The addresses in the two files starting with "matched" all have a one-to-one matching on address, even though they have street number ranges, probably because they are formatted identically. These files could be used to make a test set where the addresses should match but are not so well aligned. 

If it is possible to delineate, without too much effort, as requested above, the range start and range end in the parser, that might be useful.

**Comment by RDmitchell** [comment]
@nicholasserra -- can you investigate if this is still an issue, ie, dealing with white space in address ranges. I made a set of files that have address ranges, and do match 1:1, and the program matches them without issue on just address. They are formatted exactly the same, so maybe that is not a good test. You could potentially start with those files (in the link below, starting with "matched") and make them slightly more difficult to match, ie, one with white space and the other not. 

I believe this issue was emailed to me by another developer (even though I authored the issue, I didn't generate the content).

**Comment by RDmitchell** [comment]
Link to sample files
https://drive.google.com/open?id=0B3fTKpZ9Dx7LUlQyUG9YZkQ2aDQ

The two files with the prefix "matched" have 1:1 matching on address line 1

**Comment by nicholasserra** [comment]
Those are some nasty addresses. Anything is possible, just need to write regex to handle odd cases. Can you give me what you think the normalization output should be for those example addresses?

**Comment by RDmitchell** [comment]
Looking at some data samples (CoStar real estate data that cities import into SEED, as well as other city data), it seems like the standard might be <number><hyphen><number> without spaces, so that any hyphenated address range value would be normalized to not contain spaces before and after they hyphen for the matching comparison. And the characters in the example in this issue would get resolved to <number><hyphen><number> for the normalization and matching.

Does that make sense?

**Comment by nicholasserra** [comment]
Yeah. The more specifics, the better. Aka

This -> should be this

`123 SW Main` -> `123 sw main`
`300-22 S Green St #house:300-22` -> `300-22 s green st house 300-221`

Etc

**Comment by dnfehren-elevate** [comment]
Those were my examples based on addresses I've seen in Chicago, the # signs
might be confusing the issue here, they are showing the output of the
parser when it encounters those formats, not a part of the raw address
string. I've added output examples instead. The output below would be what
I would expect but I'm not sure what works best in terms of SEED
developement.

322 S Green St -> house_start=322, st_prefix=S, st_name=Green, st_suffix=St
300-22 S Green St -> house_start=300, house_end=322, st_prefix=S,
st_name=Green, st_suffix=St
300/322 S Green St -> house_start=300, house_end=322, st_prefix=S,
st_name=Green, st_suffix=St
300 322 S Green St -> house_start=300, house_end=322, st_prefix=S,
st_name=Green, st_suffix=St

thanks for continuing work on this!
Dan

**Comment by RDmitchell** [comment]
Thanks Dan for jumping in. I thought you were the person who was originally talking about this so thanks for clarifying. 

And addresses with spaces in the address range would get the spaces deleted during the normalization / comparison step:

300 - 302 S Green St -> 300-302 s green st

**Comment by RDmitchell** [comment]
Another wrinkle in an address, not sure how it should be interpreted in terms of matching, but found it in real data

110/120/130 First Street, SW

**Comment by pipermerriam** [comment]
@RDmitchell I've got a working version of this that covers what should be the lion's share of the address range formats.  Should catch anything with two numbers, separated by whitespace or `-,/,\` and will expand the second number if it appears to have left off a leading digit (`125-45 => 125-145`)

I looked through the linked google documents and this implementation should handle all of the cases presented in those docs.

**Comment by RDmitchell** [comment]
Great ! go ahead and do a pull request and Miles will merge your changes into develop over the weekend so we can do some testing on the seedtest instance

**Comment by pipermerriam** [comment]
PR can be found here https://github.com/SEED-platform/seed/pull/483  onto 480

**Comment by RDmitchell** [comment]
Right ! I saw it in my email queue right after I made this comment !
I am assigning this to Baptiste for testing. Thanks !!

**Comment by RDmitchell** [comment]
@bravacheLBL -- I am assigning this to you for testing next week (Nov 23) after Miles merges the code in over the weekend. Thanks !

**Comment by RDmitchell** [comment]
I am reopening this issue for Baptiste to test

**Comment by bravacheLBL** [comment]
This issue is not totally resolved in my opinion. I added a new file in Gdrive /Issue 317 (https://drive.google.com/open?id=0B3fTKpZ9Dx7LUlQyUG9YZkQ2aDQ) with the suffix "FORMAT" where I changed the format of some address line and added a field "Note" where I wrote what I changed.

I then tried to match those buildings with the other database named "CoveredBuildings" (that was imported first).

The format changes that matched correctly are:
- Deleting a dash that was separated with spaces : "1654 - 1658" -> "1654 1658"
- Uncapitalizing letters : "New York AV NW" -> "new york av nw"
- Replacing a dash separated by spaces with a slash : "1654 - 1658" -> "1654 / 1658"

The format changes that didn't match are:
- Replacing Dash + Space by Slash : "1654 - 1658" -> "1654/1658"
- Shortened second part of number : "1654 - 1658" -> "1654 - 58"
- Spelled out ST to street or NW to north-west or av to avenue
- Replaced a dash by a space: "1654-1658" -> "1654 1658" (this is different from the "deleting dash" that DID match since the initial address doesn't have spaces around in this case)

The format changes that worked once and failed the other time:
- Deleting/Adding spaces around dash : "1654 - 1658" -> "1654-1658" (or vice-versa)

I believe we can keep this issue open.
